File: lib/scripts/import_students.js
#!/usr/bin/env node
/**
 * Wellspring College Student Bulk Import Script
 * Imports students from JSS1.csv and student_credentials_clean.csv
 */

const fs = require('fs');
const https = require('https');
const readline = require('readline');

// Supabase Configuration
const SUPABASE_URL = "https://hcvyumbkonrisrxbjnst.supabase.co";
const SUPABASE_ANON_KEY = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6Imhjdnl1bWJrb25yaXNyeGJqbnN0Iiwicm9sZSI6ImFub24iLCJpYXQiOjE3NTE0NTcyOTIsImV4cCI6MjA2NzAzMzI5Mn0.6OS27VWKITYjfF5aKg7BMqxYu2wphh24O26J2-NMoew";
const SUPABASE_SERVICE_KEY = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6Imhjdnl1bWJrb25yaXNyeGJqbnN0Iiwicm9sZSI6InNlcnZpY2Vfcm9sZSIsImlhdCI6MTc1MTQ1NzI5MiwiZXhwIjoyMDY3MDMzMjkyfQ.VsoNUEhD75MOnSg5q_HKaRQ64LDapKcEbuk7vixd4mk";
const SCHOOL_ID = "3b93f5ca-3389-4285-8bb2-4248981eefe3";
const SCHOOL_NAME = "Wellspring College";

// Statistics
let successCount = 0;
let failureCount = 0;
const errors = [];

// Helper function to make HTTP requests
function makeRequest(url, options, data = null) {
  return new Promise((resolve, reject) => {
    const req = https.request(url, options, (res) => {
      let body = '';
      res.on('data', (chunk) => body += chunk);
      res.on('end', () => {
        try {
          const result = body ? JSON.parse(body) : {};
          resolve({ status: res.statusCode, data: result });
        } catch (e) {
          resolve({ status: res.statusCode, data: body });
        }
      });
    });
    
    req.on('error', reject);
    if (data) req.write(JSON.stringify(data));
    req.end();
  });
}

// Create auth user
async function createAuthUser(email, password, fullName) {
  const url = `${SUPABASE_URL}/auth/v1/signup`;
  const options = {
    method: 'POST',
    headers: {
      'apikey': SUPABASE_ANON_KEY,
      'Content-Type': 'application/json'
    }
  };
  
  const data = {
    email,
    password,
    data: { full_name: fullName }
  };
  
  try {
    const response = await makeRequest(url, options, data);
    
    if (response.status === 200 || response.status === 201) {
      const userId = response.data.user?.id;
      if (userId) {
        return { success: true, userId, error: null };
      }
      return { success: false, userId: null, error: 'No user ID in response' };
    }
    
    const errorMsg = response.data.msg || response.data.message || JSON.stringify(response.data);
    return { success: false, userId: null, error: `Auth error: ${errorMsg}` };
  } catch (e) {
    return { success: false, userId: null, error: `Exception: ${e.message}` };
  }
}

// Create profile (using service role key to bypass RLS)
async function createProfile(userId, name, schoolId, schoolName) {
  const url = `${SUPABASE_URL}/rest/v1/profiles`;
  const options = {
    method: 'POST',
    headers: {
      'apikey': SUPABASE_SERVICE_KEY,
      'Authorization': `Bearer ${SUPABASE_SERVICE_KEY}`,
      'Content-Type': 'application/json',
      'Prefer': 'return=minimal'
    }
  };
  
  const data = {
    id: userId,
    name,
    school_id: schoolId,
    school_name: schoolName,
    xp: 0,
    coins: 100,
    is_premium: false
  };
  
  try {
    const response = await makeRequest(url, options, data);
    
    if (response.status === 200 || response.status === 201) {
      return { success: true, error: null };
    }
    
    const errorMsg = response.data.message || JSON.stringify(response.data);
    return { success: false, error: `Profile error: ${errorMsg}` };
  } catch (e) {
    return { success: false, error: `Exception: ${e.message}` };
  }
}

// Import a single student
async function importStudent(email, password, fullName, grade = null) {
  console.log(`\n📝 Creating: ${email}`);
  console.log(`   Name: ${fullName}`);
  console.log(`   Grade: ${grade || 'Not specified'}`);
  
  let userId = null;
  
  // Create auth user
  const authResult = await createAuthUser(email, password, fullName);
  if (!authResult.success) {
    // Check if user already exists
    if (authResult.error.includes('already registered')) {
      console.log(`   ⚠️  User already exists, skipping to next...`);
      failureCount++;
      errors.push(`Skipped ${email}: User already registered`);
      return false;
    }
    
    failureCount++;
    const errorMsg = `Failed to create ${email}: ${authResult.error}`;
    console.log(`   ❌ ${authResult.error}`);
    errors.push(errorMsg);
    return false;
  }
  
  userId = authResult.userId;
  console.log(`   ✅ Auth user created: ${userId}`);
  
  // Small delay
  await new Promise(resolve => setTimeout(resolve, 300));
  
  // Create profile
  const profileResult = await createProfile(userId, fullName, SCHOOL_ID, SCHOOL_NAME);
  if (!profileResult.success) {
    failureCount++;
    const errorMsg = `Failed to create profile for ${email}: ${profileResult.error}`;
    console.log(`   ❌ ${profileResult.error}`);
    errors.push(errorMsg);
    return false;
  }
  
  console.log(`   ✅ Profile created`);
  successCount++;
  
  // Delay to avoid rate limiting
  await new Promise(resolve => setTimeout(resolve, 500));
  
  return true;
}

// Parse CSV
function parseCSV(content) {
  const lines = content.split('\n').filter(line => line.trim());
  const headers = lines[0].split(',').map(h => h.trim());
  const rows = [];
  
  for (let i = 1; i < lines.length; i++) {
    const values = lines[i].split(',').map(v => v.trim());
    if (values.length === headers.length) {
      const row = {};
      headers.forEach((header, index) => {
        row[header] = values[index];
      });
      rows.push(row);
    }
  }
  
  return rows;
}

// Import from JSS1.csv
async function importJSS1(filepath, testMode = false, limit = 5) {
  console.log(`\n📚 Importing from: ${filepath}`);
  if (testMode) {
    console.log(`   🧪 TEST MODE: Only importing first ${limit} students`);
  }
  
  try {
    const content = fs.readFileSync(filepath, 'utf-8');
    const rows = parseCSV(content);
    
    let count = 0;
    for (const row of rows) {
      if (testMode && count >= limit) break;
      
      const firstName = row['First Name'];
      const lastName = row['Last Name'];
      const email = row['Email Address'].toLowerCase();
      const password = row['Password'];
      
      if (!email || !password) continue;
      
      const fullName = `${firstName} ${lastName}`;
      await importStudent(email, password, fullName, 'JSS1');
      count++;
    }
    
    console.log(`✅ Processed ${count} students from JSS1.csv`);
  } catch (e) {
    console.log(`❌ Error reading JSS1.csv: ${e.message}`);
    errors.push(`JSS1 CSV error: ${e.message}`);
  }
}

// Import from student_credentials_clean.csv
async function importClean(filepath, testMode = false, limit = 5) {
  console.log(`\n📚 Importing from: ${filepath}`);
  if (testMode) {
    console.log(`   🧪 TEST MODE: Only importing first ${limit} students`);
  }
  
  try {
    const content = fs.readFileSync(filepath, 'utf-8');
    const rows = parseCSV(content);
    
    let count = 0;
    for (const row of rows) {
      if (testMode && count >= limit) break;
      
      const fullName = row['Name'];
      const email = row['Email'].toLowerCase();
      const password = row['Password'];
      
      if (!email || !password) continue;
      
      await importStudent(email, password, fullName, null);
      count++;
    }
    
    console.log(`✅ Processed ${count} students from clean CSV`);
  } catch (e) {
    console.log(`❌ Error reading clean CSV: ${e.message}`);
    errors.push(`Clean CSV error: ${e.message}`);
  }
}

// Print statistics
function printStatistics() {
  console.log('\n' + '='.repeat(60));
  console.log('📊 IMPORT STATISTICS');
  console.log('='.repeat(60));
  console.log(`✅ Successful imports: ${successCount}`);
  console.log(`❌ Failed imports: ${failureCount}`);
  console.log(`📝 Total processed: ${successCount + failureCount}`);
  
  if (errors.length > 0) {
    console.log(`\n⚠️  ERRORS (${errors.length}):`);
    errors.slice(0, 10).forEach(error => {
      console.log(`  - ${error}`);
    });
    if (errors.length > 10) {
      console.log(`  ... and ${errors.length - 10} more errors`);
    }
  }
  
  console.log('='.repeat(60) + '\n');
}

// Main function
async function main() {
  console.log('🚀 Wellspring College Student Import Script');
  console.log('='.repeat(60));
  
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
  });
  
  console.log('\nSelect import mode:');
  console.log('1. Test mode (5 students from JSS1)');
  console.log('2. Full import (all 183 students)');
  
  rl.question('\nEnter choice (1 or 2): ', async (choice) => {
    if (choice.trim() === '1') {
      // Test mode
      console.log('\n🧪 TEST MODE SELECTED');
      await importJSS1('lib/JSS1.csv', true, 5);
    } else if (choice.trim() === '2') {
      // Full import
      console.log('\n🚀 FULL IMPORT SELECTED');
      rl.question('This will import 183 students. Continue? (yes/no): ', async (confirm) => {
        if (confirm.trim().toLowerCase() === 'yes') {
          await importJSS1('lib/JSS1.csv', false);
          await importClean('lib/student_credentials_clean.csv', false);
        } else {
          console.log('Import cancelled.');
        }
        
        printStatistics();
        
        if (successCount > 0) {
          console.log('✅ Import completed successfully!');
        } else {
          console.log('❌ Import failed. Please check errors above.');
        }
        
        rl.close();
      });
      return;
    } else {
      console.log('Invalid choice. Exiting.');
      rl.close();
      return;
    }
    
    printStatistics();
    
    if (successCount > 0) {
      console.log('✅ Import completed successfully!');
    } else {
      console.log('❌ Import failed. Please check errors above.');
    }
    
    rl.close();
  });
}

main().catch(console.error);

File: lib/scripts/import_students.py
#!/usr/bin/env python3
"""
Wellspring College Student Bulk Import Script
Imports students from JSS1.csv and student_credentials_clean.csv
"""

import csv
import requests
import time
import json
from typing import Dict, List, Tuple

# Supabase Configuration
SUPABASE_URL = "https://hcvyumbkonrisrxbjnst.supabase.co"
SUPABASE_ANON_KEY = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6Imhjdnl1bWJrb25yaXNyeGJqbnN0Iiwicm9sZSI6ImFub24iLCJpYXQiOjE3NTE0NTcyOTIsImV4cCI6MjA2NzAzMzI5Mn0.6OS27VWKITYjfF5aKg7BMqxYu2wphh24O26J2-NMoew"
SCHOOL_ID = "3b93f5ca-3389-4285-8bb2-4248981eefe3"
SCHOOL_NAME = "Wellspring College"

# Statistics
success_count = 0
failure_count = 0
errors = []

def create_auth_user(email: str, password: str, full_name: str) -> Tuple[bool, str, str]:
    """Create a Supabase auth user"""
    url = f"{SUPABASE_URL}/auth/v1/signup"
    headers = {
        "apikey": SUPABASE_ANON_KEY,
        "Content-Type": "application/json"
    }
    data = {
        "email": email,
        "password": password,
        "data": {
            "full_name": full_name
        }
    }
    
    try:
        response = requests.post(url, headers=headers, json=data)
        
        if response.status_code in [200, 201]:
            result = response.json()
            user_id = result.get('user', {}).get('id')
            if user_id:
                return True, user_id, ""
            else:
                return False, "", "No user ID in response"
        else:
            error_msg = response.json().get('msg', response.text)
            return False, "", f"Auth error: {error_msg}"
    except Exception as e:
        return False, "", f"Exception: {str(e)}"

def create_profile(user_id: str, name: str, school_id: str, school_name: str) -> Tuple[bool, str]:
    """Create a profile for the user"""
    url = f"{SUPABASE_URL}/rest/v1/profiles"
    headers = {
        "apikey": SUPABASE_ANON_KEY,
        "Content-Type": "application/json",
        "Prefer": "return=minimal"
    }
    data = {
        "id": user_id,
        "name": name,
        "school_id": school_id,
        "school_name": school_name,
        "xp": 0,
        "coins": 100,
        "is_premium": False
    }
    
    try:
        response = requests.post(url, headers=headers, json=data)
        
        if response.status_code in [200, 201]:
            return True, ""
        else:
            error_msg = response.json().get('message', response.text)
            return False, f"Profile error: {error_msg}"
    except Exception as e:
        return False, f"Exception: {str(e)}"

def import_student(email: str, password: str, full_name: str, grade: str = None) -> bool:
    """Import a single student"""
    global success_count, failure_count, errors
    
    print(f"\n📝 Creating: {email}")
    print(f"   Name: {full_name}")
    print(f"   Grade: {grade or 'Not specified'}")
    
    # Create auth user
    success, user_id, error = create_auth_user(email, password, full_name)
    if not success:
        failure_count += 1
        error_msg = f"Failed to create {email}: {error}"
        print(f"   ❌ {error}")
        errors.append(error_msg)
        return False
    
    print(f"   ✅ Auth user created: {user_id}")
    
    # Small delay
    time.sleep(0.3)
    
    # Create profile
    success, error = create_profile(user_id, full_name, SCHOOL_ID, SCHOOL_NAME)
    if not success:
        failure_count += 1
        error_msg = f"Failed to create profile for {email}: {error}"
        print(f"   ❌ {error}")
        errors.append(error_msg)
        return False
    
    print(f"   ✅ Profile created")
    success_count += 1
    
    # Delay to avoid rate limiting
    time.sleep(0.5)
    
    return True

def import_jss1_csv(filepath: str, test_mode: bool = False, limit: int = 5):
    """Import students from JSS1.csv"""
    print(f"\n📚 Importing from: {filepath}")
    if test_mode:
        print(f"   🧪 TEST MODE: Only importing first {limit} students")
    
    try:
        with open(filepath, 'r', encoding='utf-8') as file:
            reader = csv.DictReader(file)
            count = 0
            
            for row in reader:
                if test_mode and count >= limit:
                    break
                
                first_name = row['First Name'].strip()
                last_name = row['Last Name'].strip()
                email = row['Email Address'].strip().lower()
                password = row['Password'].strip()
                
                if not email or not password:
                    continue
                
                full_name = f"{first_name} {last_name}"
                import_student(email, password, full_name, "JSS1")
                count += 1
        
        print(f"✅ Processed {count} students from JSS1.csv")
    except Exception as e:
        print(f"❌ Error reading JSS1.csv: {e}")
        errors.append(f"JSS1 CSV error: {e}")

def import_clean_csv(filepath: str, test_mode: bool = False, limit: int = 5):
    """Import students from student_credentials_clean.csv"""
    print(f"\n📚 Importing from: {filepath}")
    if test_mode:
        print(f"   🧪 TEST MODE: Only importing first {limit} students")
    
    try:
        with open(filepath, 'r', encoding='utf-8') as file:
            reader = csv.DictReader(file)
            count = 0
            
            for row in reader:
                if test_mode and count >= limit:
                    break
                
                full_name = row['Name'].strip()
                email = row['Email'].strip().lower()
                password = row['Password'].strip()
                
                if not email or not password:
                    continue
                
                import_student(email, password, full_name, None)
                count += 1
        
        print(f"✅ Processed {count} students from clean CSV")
    except Exception as e:
        print(f"❌ Error reading clean CSV: {e}")
        errors.append(f"Clean CSV error: {e}")

def print_statistics():
    """Print final statistics"""
    print("\n" + "=" * 60)
    print("📊 IMPORT STATISTICS")
    print("=" * 60)
    print(f"✅ Successful imports: {success_count}")
    print(f"❌ Failed imports: {failure_count}")
    print(f"📝 Total processed: {success_count + failure_count}")
    
    if errors:
        print(f"\n⚠️  ERRORS ({len(errors)}):")
        for error in errors[:10]:  # Show first 10 errors
            print(f"  - {error}")
        if len(errors) > 10:
            print(f"  ... and {len(errors) - 10} more errors")
    
    print("=" * 60 + "\n")

def main():
    """Main execution"""
    print("🚀 Wellspring College Student Import Script")
    print("=" * 60)
    
    # Ask user for mode
    print("\nSelect import mode:")
    print("1. Test mode (5 students from JSS1)")
    print("2. Full import (all 183 students)")
    
    choice = input("\nEnter choice (1 or 2): ").strip()
    
    if choice == "1":
        # Test mode
        print("\n🧪 TEST MODE SELECTED")
        import_jss1_csv('../JSS1.csv', test_mode=True, limit=5)
    elif choice == "2":
        # Full import
        print("\n🚀 FULL IMPORT SELECTED")
        confirm = input("This will import 183 students. Continue? (yes/no): ").strip().lower()
        if confirm == "yes":
            import_jss1_csv('../JSS1.csv', test_mode=False)
            import_clean_csv('../student_credentials_clean.csv', test_mode=False)
        else:
            print("Import cancelled.")
            return
    else:
        print("Invalid choice. Exiting.")
        return
    
    print_statistics()
    
    if success_count > 0:
        print("✅ Import completed successfully!")
    else:
        print("❌ Import failed. Please check errors above.")

if __name__ == "__main__":
    main()
